cbmc_viewer/memopt.py

```diff
@@ -0,0 +1,204 @@
+
+    def __repr__(self):
+        """A dict representation of memop summary."""
+
+        self.validate()
+        return self.__dict__
+
+    def __str__(self):
+        """A string representation of memop summary."""
+
+        return json.dumps({JSON_TAG: self.__repr__()}, indent=2)
+
+    def validate(self):
+        """Validate members of a summary object."""
+
+        return voluptuous.humanize.validate_with_humanized_errors(
+            self.__dict__, VALID_MEMOP_SUMMARY
+        )
+
+    def dump(self, filename=None, outdir=None):
+        """Write memop metrics to a file or stdout."""
+
+        util.dump(self, filename, outdir)
+
+    def render_report(self, filename=None, outdir=None):
+        """Write the memop summary to a file rendered as html."""
+
+        memop_html = templates.render_memop_summary(self.summary)
+        util.dump(memop_html, filename or "memop.html", outdir)
+
+################################################################
+# Json key tags to access elements in cbmc json output
+
+JSON_FUNCTIONS_KEY = 'functions'
+JSON_BODY_AVAIL_KEY = 'isBodyAvailable'
+JSON_INSTRUCTIONS_KEY = 'instructions'
+JSON_INSTRUCTION_ID_KEY = 'instructionId'
+JSON_OPERANDS_KEY = 'operands'
+JSON_FUNCTION_INSTR_KEY = 'instruction'
+JSON_SOURCE_LOC_KEY = 'sourceLocation'
+
+# Json keys for CBMC operands
+JSON_ID_KEY = 'id'
+JSON_NAMED_SUB_KEY = 'namedSub'
+JSON_IDENTIFIER_KEY = 'identifier'
+
+# List of memory operations
+list_mem_func = ["memcmp","memcpy","memmove","memset","memccpy","memchr"]
+
+# Example cbmc json output
+# [
+#   {
+#     "program": "CBMC 5.13.0 ..."
+#   },
+#   ...
+#   {
+#     "functions": [
+#       {
+#         "instructions": [
+#           {
+#             "guard": {
+#               "id": "nil"
+#             },
+#             "instruction": "... struct.c line 23...\n...memcmp...\n",
+#             "instructionId": "FUNCTION_CALL",
+#             "operands": [
+#               ...
+#               {
+#                 "id": "symbol",
+#                 "namedSub": {
+#                   "identifier": {
+#                     "id": "memcmp"
+#                   },
+#                   ...
+#                 }
+#               },
+#               ...
+#             ],
+#             "sourceLocation": {
+#               "file": "struct.c",
+#               "function": "main",
+#               "line": "23",
+#               "workingDirectory": "..."
+#             }
+#           },
+#           ...
+#         ],
+#         "isBodyAvailable": true,
+#         ...
+#       }
+#     ]
+#   }
+# ]
+
+################################################################
+# Utility functions to generate memory op call summary
+
+def is_memop_call(operands):
+    for item in operands:
+        if item[JSON_ID_KEY] == "symbol":
+            func_name = \
+                item[JSON_NAMED_SUB_KEY][JSON_IDENTIFIER_KEY][JSON_ID_KEY]
+            return any([True for func in list_mem_func if func in func_name])
```

Comment:
Is this really doing what you intend?  The test func in func_name is testing whether the string func is a substring of func_name.  Is that what you want?  Wouldn't it be simpler to write `any([func in func_name for func in list_mem_func])`

Reply:
Yes we want to see if `memcpy`, `memset` etc are substrings of the func_name.
Yeah this would be simpler, I have made the change.

Reply:
What we are extracting is the entire function call memcpy(...) or __builtin_memcpy(...).  Also tracks stubs.